### PR TITLE
[dotnetcore] Implement QueryParameter deepObject style is missing for…

### PR DESCRIPTION
**Pull request by MachineUserPTV**
[dotnetcore] Implement QueryParameter deepObject style for client generators dotnetcore for async functions.

@mandrean (2017/08) @frankyjuang (2019/09) @shibayan (2020/02) @wing328 

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [ ] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [ ] In case you are adding a new generator, run the following additional script : 
  ```
  ./bin/utils/ensure-up-to-date.sh
  ``` 
  Commit all changed files.
- [x] File the PR against the [correct branch] : master. 6.6.x
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

Fix for 
https://github.com/OpenAPITools/openapi-generator/issues/14977

This PR replaces the old PR 14979:
https://github.com/OpenAPITools/openapi-generator/pull/14979/
 - I don't know what went wrong with the old one.

### CODE REVIEW

**Comment by wing328** [comment]
@MachineUserPTV thanks for the PR.

Would it be possible to add a test?

We add tests for deepObject style using the echo api spec.

You can copy `./bin/configs/python-echo-api.yaml` to create `./bin/configs/csharp-echo-api.yaml` as a starting point.

**Comment by wing328** [comment]
I tested it locally but got the following errors:
```
ExecutionThreadApartmentState option not supported for framework: .NETCoreApp,Version=v6.0.
[xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v2.5.0.1+5ebf84cd75 (64-bit .NET 6.0.16)
[xUnit.net 00:00:00.12]   Discovering: Org.OpenAPITools.Test
[xUnit.net 00:00:00.18]   Discovered:  Org.OpenAPITools.Test
[xUnit.net 00:00:00.18]   Starting:    Org.OpenAPITools.Test
[xUnit.net 00:00:02.98]     Org.OpenAPITools.Test.Api.BodyApiTests.TestEchoBodyPetTest [FAIL]
[xUnit.net 00:00:02.98]       Assert.IsType() Failure: Value is null
[xUnit.net 00:00:02.98]       Expected: typeof(Org.OpenAPITools.Model.Pet)
[xUnit.net 00:00:02.98]       Actual:   null
[xUnit.net 00:00:02.98]       Stack Trace:
[xUnit.net 00:00:02.99]         /Users/williamcheng/Code/openapi-generator4/samples/client/echo_api/csharp-restsharp/src/Org.OpenAPITools.Test/Api/BodyApiTests.cs(102,0): at Org.OpenAPITools.Test.Api.BodyApiTests.TestEchoBodyPetTest()
Running Org.OpenAPITools.Org.OpenAPITools.Test.Org.OpenAPITools.Test.Api.BodyApiTests.TestEchoBodyPetTest ...
Assert.IsType() Failure: Value is null
Expected: typeof(Org.OpenAPITools.Model.Pet)
Actual:   nullResult message:
Assert.IsType() Failure: Value is null
Expected: typeof(Org.OpenAPITools.Model.Pet)
Actual:   null[xUnit.net 00:00:08.66]   Finished:    Org.OpenAPITools.Test

```

Do the tests pass locally in your machine?

**Comment by wing328** [comment]
After updating the samples, I got more errors:
```
ExecutionThreadApartmentState option not supported for framework: .NETCoreApp,Version=v6.0.
[xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v2.5.0.1+5ebf84cd75 (64-bit .NET 6.0.16)
[xUnit.net 00:00:00.14]   Discovering: Org.OpenAPITools.Test
[xUnit.net 00:00:00.21]   Discovered:  Org.OpenAPITools.Test
[xUnit.net 00:00:00.22]   Starting:    Org.OpenAPITools.Test
[xUnit.net 00:00:02.68]     Org.OpenAPITools.Test.Api.BodyApiTests.TestEchoBodyPetTest [FAIL]
[xUnit.net 00:00:02.68]       Assert.IsType() Failure: Value is null
[xUnit.net 00:00:02.68]       Expected: typeof(Org.OpenAPITools.Model.Pet)
[xUnit.net 00:00:02.69]       Actual:   null
[xUnit.net 00:00:02.69]       Stack Trace:
[xUnit.net 00:00:02.69]         /Users/williamcheng/Code/openapi-generator4/samples/client/echo_api/csharp-restsharp/src/Org.OpenAPITools.Test/Api/BodyApiTests.cs(102,0): at Org.OpenAPITools.Test.Api.BodyApiTests.TestEchoBodyPetTest()
Running Org.OpenAPITools.Org.OpenAPITools.Test.Org.OpenAPITools.Test.Api.BodyApiTests.TestEchoBodyPetTest ...
Assert.IsType() Failure: Value is null
Expected: typeof(Org.OpenAPITools.Model.Pet)
Actual:   nullResult message:
Assert.IsType() Failure: Value is null
Expected: typeof(Org.OpenAPITools.Model.Pet)
Actual:   null[xUnit.net 00:00:08.74]     Org.OpenAPITools.Test.CustomTest.testQueryStyleDeepObjectExplodeTrueObjectAsyncTest [FAIL]
[xUnit.net 00:00:08.74]       Assert.Equal() Failure: Strings differ
[xUnit.net 00:00:08.74]                                         ↓ (pos 49)
[xUnit.net 00:00:08.74]       Expected: ···"de_true/object?query_object=class%20Pet%2"···
[xUnit.net 00:00:08.74]       Actual:   ···"de_true/object?queryObject%5bid%5d=12345&"···
[xUnit.net 00:00:08.74]                                         ↑ (pos 49)
[xUnit.net 00:00:08.74]       Stack Trace:
[xUnit.net 00:00:08.74]         /Users/williamcheng/Code/openapi-generator4/samples/client/echo_api/csharp-restsharp/src/Org.OpenAPITools.Test/CustomTest.cs(48,0): at Org.OpenAPITools.Test.CustomTest.testQueryStyleDeepObjectExplodeTrueObjectAsyncTest()
[xUnit.net 00:00:08.74]   Finished:    Org.OpenAPITools.Test
Running Org.OpenAPITools.Org.OpenAPITools.Test.Org.OpenAPITools.Test.CustomTest.testQueryStyleDeepObjectExplodeTrueObjectAsyncTest ...
Assert.Equal() Failure: Strings differ
                                  ↓ (pos 49)
Expected: ···"de_true/object?query_object=class%20Pet%2"···
Actual:   ···"de_true/object?queryObject%5bid%5d=12345&"···
                                  ↑ (pos 49)Result message:
Assert.Equal() Failure: Strings differ
                                  ↓ (pos 49)
Expected: ···"de_true/object?query_object=class%20Pet%2"···
Actual:   ···"de_true/object?queryObject%5bid%5d=12345&"···
                                  ↑ (pos 49)
```

**Comment by wing328** [comment]
let me know if you need help updating the samples: https://github.com/OpenAPITools/openapi-generator/actions/runs/6194699000/job/16818555563?pr=15945

File: bin/configs/go-echo-server-petstore-new.yaml [hunk 1]
```diff
@@ -1,9 +0,0 @@
-generatorName: go-echo-server
-outputDir: samples/server/petstore/go-echo-server
-inputSpec: modules/openapi-generator/src/test/resources/3_0/petstore.yaml
-templateDir: modules/openapi-generator/src/main/resources/go-echo-server
-gitHost: gitlab.com
-gitUserId: openapitools
-gitRepoId: petstore
-additionalProperties:
-  hideGenerationTimestamp: "true"
```
Comment (wing328) [inline bin/configs/go-echo-server-petstore-new.yaml:9]: looks like these files are accidentally removed. can you please restore them?

**Comment by wing328** [comment]
Tested with the latest and got a different error:
```
[xUnit.net 00:00:10.67]         /Users/williamcheng/Code/openapi-generator4/samples/client/echo_api/csharp-restsharp/src/Org.OpenAPITools.Test/CustomTest.cs(64,0): at Org.OpenAPITools.Test.CustomTest.TestQueryStyleFormExplodeTrueObjectAllOfTest()
Running Org.OpenAPITools.Org.OpenAPITools.Test.Org.OpenAPITools.Test.CustomTest.TestQueryStyleFormExplodeTrueObjectAllOfTest ...
Assert.Equal() Failure: Strings differ
                                  ↓ (pos 341)
Expected: ···"d%0a%20%20Date%3a%2001.01.0001%2000%3a00%"···
Actual:   ···"d%0a%20%20Date%3a%201%2f1%2f0001%2012%3a0"···
                                  ↑ (pos 341)Result message:
Assert.Equal() Failure: Strings differ
                                  ↓ (pos 341)
Expected: ···"d%0a%20%20Date%3a%2001.01.0001%2000%3a00%"···
Actual:   ···"d%0a%20%20Date%3a%201%2f1%2f0001%2012%3a0"···
                                  ↑ (pos 341)[xUnit.net 00:00:12.68]   Finished:    Org.OpenAPITools.Test
```

Did the tests passed locally without issue in your side?

**Comment by MachineUserPTV** [comment]
I thought they worked ? But now they didn't. I fixed them.

**Comment by wing328** [comment]
Tested locally and all tests passed. Thanks for the PR.

**Comment by wing328** [comment]
FYI. Filed https://github.com/OpenAPITools/openapi-generator/pull/16626 with minor enhancement

**Pull request merged**